awx/main/scheduler/task_manager.py

```diff
@@ -452,56 +447,20 @@ def generate_dependencies(self, task):
 
             if len(dependencies) > 0:
                 self.capture_chain_failure_dependencies(task, dependencies)
-        return dependencies
-
-    def process_dependencies(self, dependent_task, dependency_tasks):
-        for task in dependency_tasks:
-            if self.is_job_blocked(task):
-                logger.debug("Dependent {} is blocked from running".format(task.log_format))
-                continue
-            preferred_instance_groups = task.preferred_instance_groups
-            found_acceptable_queue = False
-            idle_instance_that_fits = None
-            for rampart_group in preferred_instance_groups:
-                if idle_instance_that_fits is None:
-                    idle_instance_that_fits = rampart_group.find_largest_idle_instance()
-                if not rampart_group.is_containerized and self.get_remaining_capacity(rampart_group.name) <= 0:
-                    logger.debug("Skipping group {} capacity <= 0".format(rampart_group.name))
-                    continue
-
-                execution_instance = rampart_group.fit_task_to_most_remaining_capacity_instance(task)
-                if execution_instance:
-                    logger.debug("Starting dependent {} in group {} instance {}".format(
-                                 task.log_format, rampart_group.name, execution_instance.hostname))
-                elif not execution_instance and idle_instance_that_fits:
-                    if not rampart_group.is_containerized:
-                        execution_instance = idle_instance_that_fits
-                        logger.debug("Starting dependent {} in group {} on idle instance {}".format(
-                                     task.log_format, rampart_group.name, execution_instance.hostname))
-                if execution_instance or rampart_group.is_containerized:
-                    self.graph[rampart_group.name]['graph'].add_job(task)
-                    tasks_to_fail = [t for t in dependency_tasks if t != task]
-                    tasks_to_fail += [dependent_task]
-                    self.start_task(task, rampart_group, tasks_to_fail, execution_instance)
-                    found_acceptable_queue = True
-                    break
-                else:
-                    logger.debug("No instance available in group {} to run job {} w/ capacity requirement {}".format(
-                                 rampart_group.name, task.log_format, task.task_impact))
-            if not found_acceptable_queue:
-                logger.debug("Dependent {} couldn't be scheduled on graph, waiting for next cycle".format(task.log_format))
 
     def process_pending_tasks(self, pending_tasks):
         running_workflow_templates = set([wf.unified_job_template_id for wf in self.get_running_workflow_jobs()])
         for task in pending_tasks:
-            self.process_dependencies(task, self.generate_dependencies(task))
+            if not task.dependent_jobs.exists():
```

Comment:
Does this result in an sql query?

Reply:
yes it does -- a little faster than task.dependent_jobs.all() though

I've also written a task manager version that adds a "dependencies_processed" boolean field to UnifiedJob, as a way to keeping track if dependencies have been generated. It does offer a small performance boost (x8 faster in 5k benchmark), but I figured I'd submit a more simple version first.

It can be viewed here https://github.com/ansible/awx/compare/devel...fosterseth:tm_processed_field

Reply:
I really like the idea of adding the field `dependencies_processed`. While efficiency is a factor, it would also be a heck of a lot easier to read. The check here `if not task.dependent_jobs.exists()` is actually kind of confusing to me, because many jobs do not have dependencies, so those 2 scenarios are muddled together. But if you did add it, I would want to be very clear about what values should be expected for jobs without dependencies. You might even consider setting the field to `True` in `UnifiedJob.create_unified_job` if it does not require any dependencies.